src/main/java/org/opentestsystem/ap/migration/migration/MigrationQ168.java

```
@@ -0,0 +1,232 @@
+
+/**
+ * This migration expands the GI archive zip and:
+ * 1. Ingests the GAX contents into the item json
+ * 2. Extracts and uploads GAX-referenced resources to S3
+ * 3. Adds Attachment references for the GAX-referenced resources to the item json
+ *
+ * NOTE: This migration also touches every Item json to perform model version migration
+ * via the new {@link org.opentestsystem.ap.common.model.versioning.ItemDeserializer}
+ *
+ * This has to handle a number of states:
+ * 1. The item is not being edited (no "core" branch edits)
+ * 2. The item is being edited, but a new archive has not been uploaded
+ *   (In this case, the "master" archive is downloaded, expanded, and
+ *    resources are uploaded into the editing branch)
+ * 3. The item is being edited, and a new archive has been uploaded
+ *   (In this case, the "core" archive is downloaded, expanded, and
+ *    resources are uploaded into the editing branch)
+ *
+ * NOTE: This implementation is heavily based on the IMS ArchiveService#saveGiArchive method
+ */
+@Component
+@Slf4j
+public class MigrationQ168 extends AbstractMigration {
+    private static final Pattern TRANSLATED_RESOURCE_FILENAME = Pattern.compile("^(.+)_(.+?)$");
+
+    protected MigrationQ168(final ApplicationDependencyProvider applicationDependencyProvider, final ApplicationProperties applicationProperties, final DataStoreDataManager dataManager, final DataStoreItemManager dataStoreItemManager, final DataStoreUtility dataStoreUtility, final DataStoreAttachmentManager dataStoreAttachmentManager) {
+        super(applicationDependencyProvider, applicationProperties, dataManager, dataStoreItemManager, dataStoreUtility, dataStoreAttachmentManager);
+    }
+
+    @Override
+    protected ItemEntity migrateEntity(final ItemEntity itemEntity, final MigrationContext migrationContext) {
+        if (!TYPE_GI.equalsIgnoreCase(itemEntity.getItemJson().getType())) {
+            return itemEntity;
+        }
```

Here is another area I admit some confusion.  I think we want to throw SkipMigration here instead of returning the entity.  @millerts  - Is that correct?  I recall you mentioning this recently.

---

@bdrainer , yes, you would normally only want to migrate items you're touching with the migration.  However, this is a migration that needs to touch ever item to make sure that all items have the latest model version set properly.

Another approach is that Will leaves this one only impacting GI and we create another one that simply "touches" the item.

---

I'm planning on leaving this touching/updating every item unless I hear otherwise?

---

Sounds good.